src/test_application/activescenes/scenarios_enginetest.cpp

```diff
@@ -0,0 +1,317 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "scenarios.h"
+
+#include "../ActiveApplication.h"
+
+#include <osp/Active/basic.h>
+#include <osp/Active/drawing.h>
+
+#include <osp/Active/SysHierarchy.h>
+
+#include <osp/Active/SysRender.h>
+#include <osp/Active/opengl/SysRenderGL.h>
+
+#include <osp/Shaders/Phong.h>
+#include <osp/Shaders/MeshVisualizer.h>
+
+#include <osp/id_registry.h>
+#include <osp/Resource/Package.h>
+
+#include <Magnum/Trade/MeshData.h>
+#include <Magnum/Trade/ImageData.h>
+
+#include <Magnum/GL/DefaultFramebuffer.h>
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+using Magnum::Trade::MeshData;
+using Magnum::Trade::ImageData2D;
+
+// for the 0xrrggbb_rgbf and angle literals
+using namespace Magnum::Math::Literals;
+
+namespace testapp::enginetest
+{
+
+// Materials used by the test scene. A more general application may want to
+// generate IDs at runtime, and map them to named identifiers.
+constexpr int const gc_mat_common      = 0;
+constexpr int const gc_mat_visualizer  = 1;
+
+constexpr int const gc_maxMaterials = 2;
+
+/**
+ * @brief State of the entire engine test scene
+ */
+struct EngineTestScene
+{
+    // ID registry generates entity IDs, and keeps track of which ones exist
+    osp::IdRegistry<osp::active::ActiveEnt> m_activeIds;
```

Comment:
Does this actually completely replace the EnTT entity system? I suppose most of the ent-related functions we're used to are members of the component pools. Does this change preclude us from using functions like `reg.all_of<T...>(entity)` to check for component existence, or `reg.valid(entity)`? Are there replacements for these functionalities?

Reply:
As far as I'm concerned, this works as a more capable replacement. Memory usage is lower, more efficient recycling of deleted entity IDs, but the main benefits come from the flexibility of manually managing component pools. The only thing we lose is the existing automation previously provided by the registry (and also entity versions, which we can implement if we really want to). From working on other branches, passing component pools as function parameters eliminates the need for many of the convenient functions in the first place.

`reg.all_of<>` on the inside is calling contains() on each sparse set in the registry. There aren't any instances of all_of being called with any more than a single component in the codebase. `reg.valid(ent)` is roughly equivalent to `m_activeIds.exists(ent)`.